src/samplers/orthogonal.cpp

```diff
@@ -0,0 +1,225 @@
+            Log(Warn, "Sample count should be the square of a prime"
+                "number, rounding to %i", sqr(m_resolution));
+
+        m_sample_count = sqr(m_resolution);
+    }
+
+    ref<Sampler<Float, Spectrum>> clone() override {
+        OrthogonalSampler *sampler       = new OrthogonalSampler();
+        sampler->m_jitter                = m_jitter;
+        sampler->m_strength              = m_strength;
+        sampler->m_sample_count          = m_sample_count;
+        sampler->m_resolution            = m_resolution;
+        sampler->m_samples_per_wavefront = m_samples_per_wavefront;
+        sampler->m_base_seed             = m_base_seed;
+        return sampler;
+    }
+
+    void seed(uint64_t seed_offset, size_t wavefront_size) override {
+        Base::seed(seed_offset, wavefront_size);
+        m_permutation_seed = compute_per_sequence_seed(seed_offset);
+    }
+
+    Float next_1d(Mask active = true) override {
+        if (unlikely(m_strength != 2)) {
+            return bush(current_sample_index(),
+                        m_dimension_index++,
+                        m_resolution,
+                        m_strength,
+                        m_permutation_seed, active);
+        } else {
+            return bose(current_sample_index(),
+                        m_dimension_index++,
+                        m_resolution,
+                        m_permutation_seed, active);
+        }
+    }
+
+    Point2f next_2d(Mask active = true) override {
+        Float f1 = next_1d(active),
+              f2 = next_1d(active);
+        return Point2f(f1, f2);
+    }
+
+    std::string to_string() const override {
+        std::ostringstream oss;
+        oss << "OrthogonalSampler[" << std::endl
+            << "  sample_count = " << m_sample_count << std::endl
+            << "  jitter = " << m_jitter << std::endl
+            << "]";
+        return oss.str();
+    }
+
+    MTS_DECLARE_CLASS()
+
+private:
+    /// Compute the digits of decimal value \ref i expressed in base \ref b
+    std::vector<UInt32> to_base_s(UInt32 i, ScalarUInt32 b, ScalarUInt32 t) {
+        std::vector<UInt32> digits(t);
+        for (size_t ii = 0; ii < t; i /= b, ++ii)
+            digits[ii] = i % b;
+        return digits;
+    }
+
+    /// Evaluate polynomial with coefficients \ref x at location arg
+    UInt32 eval_poly(const std::vector<UInt32> &coef, UInt32 x) {
+        UInt32 res = 0;
+        for (size_t l = coef.size(); l--; )
+            res = (res * x) + coef[l];
+        return res;
+    }
+
+    /// Bush construction technique for orthogonal arrays
+    Float bush(UInt32 i,       // sample index
+               ScalarUInt32 j, // dimension
+               ScalarUInt32 s, // number of levels/stratas (has to be a prime number)
+               ScalarUInt32 t, // strength of OA
+               UInt32 p,       // pseudo-random permutation seed
+               Mask active = true) {
+        ScalarUInt32 N = enoki::pow(s, t);
```

Comment:
`ScalarUInt32` -> `uint32_t` everywhere.

Normally `pow(a, b) = exp(log(a) * b)`, which would be a very bad idea. There is a special case somewhere in Enoki that specializes to integer exponent. It would be good to double-check that that is executed here..